### PR TITLE
lua: implement stubs for etcd client

### PR DESCRIPTION
**Pull request by ligurio**
lua: implement stubs for etcd client
    
    To enable etcd client set the following CMake variables:
    
     - ENABLE_ETCD_CLIENT: ON
     - ETCD_CLIENT_DIR: path to a directory with etcd client Lua source
       files
    
    If ENABLE_ETCD_CLIENT is unset, then a stub implementation is built,
    which spits an error on attempt require etcd-client.
    
    Needed for https://github.com/tarantool/tarantool-ee/issues/443
    
    MSG="stubs for enterprise version, no user-visible changes"
    
    NO_CHANGELOG=MSG
    NO_DOC=MSG
    NO_TEST=MSG

Needed for https://github.com/tarantool/tarantool-ee/issues/443

### CODE REVIEW

File: src/box/CMakeLists.txt [hunk 1]
```diff
@@ -29,6 +29,23 @@ endif()
...
+    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/grpc_json.lua etcd_client_grpc_json_lua)
+    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/protocol.lua etcd_client_protocol_lua)
+    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/subscribe/auto.lua etcd_client_subscribe_auto_lua)
+    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/subscribe/once.lua etcd_client_subscribe_once_lua)
+    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/utils.lua etcd_client_utils_lua)
+    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/init.lua etcd_client_init_lua)
+    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/client/pool.lua etcd_client_client_pool_lua)
+    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/client/jsstream.lua etcd_client_client_jsstream_lua)
+    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/client/http.lua etcd_client_client_http_lua)
+    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/client/auth_http.lua etcd_client_client_auth_http_lua)
```
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/box/CMakeLists.txt:47]: On 2: it is not a third-party module, it is our own one.

[the remark this answers is not in the document]
Comment (locker) [reply src/box/CMakeLists.txt:47]: `metrics` and `checks` are also our own modules, still they're located in `third_party`. Let's be consistent. From the `tarantool-ee` point of view, `etcd-client` is third-party code.

[the remark this answers is not in the document]
Comment (Totktonada) [reply src/box/CMakeLists.txt:47]: I think that `metrics` and `checks` should be in `src/lua` (maybe in `src/box/lua`) as well.

I see the draw line: `src/lib` has our own code (including one that is developed in its own repository), `third_party` contains externally originated code. It seems logical to apply the same rules for Lua code.

As regarding the linter checks: if it is already checked in the module repository, tarantool's repository check shouldn't reveal anything new. Well, we can save 1 second of machine time, but meaningful file tree structure looks more important.

TBH, I don't want to be mad on insisting here. I just see no logic in placing some our code in `src`, but some in `third_party`. I can accept the 'consistency' point (which actually brings more inconsistency, IMHO).

[the remark this answers is not in the document]
Comment (locker) [reply src/box/CMakeLists.txt:47]: Discussed with @alyapunov. Agreed that we should store this in src/lua/etcd-client. Please ignore point 2. We should also move checks and metrics submodules to src/lua or src/box/lua.

[the remark this answers is not in the document]
Comment (ligurio) [reply src/box/CMakeLists.txt:47]: Fixed pp. 1, 2:

```diff
diff --git a/src/box/CMakeLists.txt b/src/box/CMakeLists.txt
index caea234a7..3c6330a61 100644
--- a/src/box/CMakeLists.txt
+++ b/src/box/CMakeLists.txt
@@ -30,21 +30,21 @@ if(ENABLE_SECURITY)
     lua_source(lua_sources ${SECURITY_LUA_SOURCE} security_lua)
 endif()
 if(ENABLE_ETCD_CLIENT)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/error/etcd.lua etcd_client_error_etcd_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/error/http.lua etcd_client_error_http_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/txn.lua etcd_client_txn_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/version.lua etcd_client_version_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/math.lua etcd_client_math_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/grpc_json.lua etcd_client_grpc_json_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/protocol.lua etcd_client_protocol_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/subscribe/auto.lua etcd_client_subscribe_auto_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/subscribe/once.lua etcd_client_subscribe_once_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/utils.lua etcd_client_utils_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/init.lua etcd_client_init_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/client/pool.lua etcd_client_client_pool_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/client/jsstream.lua etcd_client_client_jsstream_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/client/http.lua etcd_client_client_http_lua)
-    lua_source(lua_sources ${EE_ROOT_DIR}/src/lua/etcd-client/etcd-client/client/auth_http.lua etcd_client_client_auth_http_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/client/auth_http.lua etcd_client_client_auth_http_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/client/http.lua etcd_client_client_http_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/client/jsstream.lua etcd_client_client_jsstream_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/client/pool.lua etcd_client_client_pool_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/error/etcd.lua etcd_client_error_etcd_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/error/http.lua etcd_client_error_http_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/grpc_json.lua etcd_client_grpc_json_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/init.lua etcd_client_init_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/math.lua etcd_client_math_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/protocol.lua etcd_client_protocol_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/subscribe/auto.lua etcd_client_subscribe_auto_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/subscribe/once.lua etcd_client_subscribe_once_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/txn.lua etcd_client_txn_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/utils.lua etcd_client_utils_lua)
+    lua_source(lua_sources ${ETCD_CLIENT_DIR}/etcd-client/version.lua etcd_client_version_lua)
 endif()
 lua_source(lua_sources lua/net_box.lua net_box_lua)
 lua_source(lua_sources lua/upgrade.lua upgrade_lua)
```


```diff
 #if ENABLE_ETCD_CLIENT
+       etcd_client_client_auth_http_lua[],
+       etcd_client_client_http_lua[],
+       etcd_client_client_jsstream_lua[],
+       etcd_client_client_pool_lua[],
        etcd_client_error_etcd_lua[],
        etcd_client_error_http_lua[],
-       etcd_client_txn_lua[],
-       etcd_client_version_lua[],
-       etcd_client_math_lua[],
        etcd_client_grpc_json_lua[],
+       etcd_client_init_lua[],
+       etcd_client_math_lua[],
        etcd_client_protocol_lua[],
        etcd_client_subscribe_auto_lua[],
        etcd_client_subscribe_once_lua[],
+       etcd_client_txn_lua[],
        etcd_client_utils_lua[],
-       etcd_client_init_lua[],
-       etcd_client_client_pool_lua[],
-       etcd_client_client_jsstream_lua[],
-       etcd_client_client_http_lua[],
-       etcd_client_client_auth_http_lua[],
+       etcd_client_version_lua[],
 #endif
```

```diff
@@ -248,6 +248,10 @@ static const char *lua_sources[] = {
        "box/key_def", "key_def", key_def_lua,
        "box/merger", "merger", merger_lua,
 #if ENABLE_ETCD_CLIENT
+       /*
+        * Module components order is important here: components that required
+        * other modules must be loaded first.
+        */
        "lua/etcd-client/etcd-client/version",
        "etcd-client.version", etcd_client_version_lua,
        "lua/etcd-client/etcd-client/utils",
```

File: src/box/lua/init.c [hunk 2]
```diff
@@ -230,6 +247,42 @@ static const char *lua_sources[] = {
...
+	"etcd-client.client.pool", etcd_client_client_pool_lua,
+	"lua/etcd-client/etcd-client/client/jsstream",
+	"etcd-client.client.jsstream", etcd_client_client_jsstream_lua,
+	"lua/etcd-client/etcd-client/client/http",
+	"etcd-client.client.http", etcd_client_client_http_lua,
+	"lua/etcd-client/etcd-client/client/auth_http",
+	"etcd-client.client.auth_http", etcd_client_client_auth_http_lua,
+	"lua/etcd-client/etcd-client/init",
+	"etcd-client", etcd_client_init_lua,
+#endif
```
[the remark this answers is not in the document]
Comment (ligurio) [reply src/box/lua/init.c:285]: Sure, I'll move.

[the remark this answers is not in the document]
Comment (ligurio) [reply src/box/lua/init.c:285]: sources moved to src/lua, commit message has been updated

**Review by locker — APPROVED** [review]

**Pull request merged**